File: python/backtest.py
"""
Backtesting framework for ELECTRA-based trading strategies.

This module provides tools for backtesting trading strategies that use
ELECTRA sentiment analysis for signal generation and position sizing.
"""

import torch
import torch.nn as nn
import numpy as np
import pandas as pd
from typing import Optional, List, Dict, Tuple
from dataclasses import dataclass

from .model import FinancialElectra, ElectraTradingSignal


@dataclass
class Trade:
    """Represents a single trade."""
    entry_date: pd.Timestamp
    exit_date: pd.Timestamp
    direction: int
    entry_price: float
    exit_price: float
    position_size: float
    pnl: float
    pnl_pct: float


@dataclass
class BacktestResult:
    """Results from a backtest."""
    total_return: float
    sharpe_ratio: float
    sortino_ratio: float
    max_drawdown: float
    win_rate: float
    profit_factor: float
    num_trades: int
    trades: List[Trade]
    equity_curve: pd.Series
    daily_returns: pd.Series


def compute_sharpe_ratio(
    returns: pd.Series,
    risk_free_rate: float = 0.02,
    periods_per_year: int = 252
) -> float:
    """
    Compute annualized Sharpe ratio.

    Args:
        returns: Series of daily returns
        risk_free_rate: Annual risk-free rate
        periods_per_year: Number of trading periods per year

    Returns:
        Annualized Sharpe ratio
    """
    if returns.std() == 0:
        return 0.0

    daily_rf = risk_free_rate / periods_per_year
    excess_returns = returns - daily_rf
    return float(np.sqrt(periods_per_year) * excess_returns.mean() / excess_returns.std())


def compute_sortino_ratio(
    returns: pd.Series,
    risk_free_rate: float = 0.02,
    periods_per_year: int = 252
) -> float:
    """
    Compute annualized Sortino ratio.

    Args:
        returns: Series of daily returns
        risk_free_rate: Annual risk-free rate
        periods_per_year: Number of trading periods per year

    Returns:
        Annualized Sortino ratio
    """
    daily_rf = risk_free_rate / periods_per_year
    excess_returns = returns - daily_rf
    downside_returns = excess_returns[excess_returns < 0]

    if len(downside_returns) == 0 or downside_returns.std() == 0:
        return 0.0

    return float(np.sqrt(periods_per_year) * excess_returns.mean() / downside_returns.std())


def compute_max_drawdown(equity_curve: pd.Series) -> float:
    """
    Compute maximum drawdown from an equity curve.

    Args:
        equity_curve: Series of portfolio values

    Returns:
        Maximum drawdown as a negative fraction
    """
    peak = equity_curve.expanding().max()
    drawdown = (equity_curve - peak) / peak
    return float(drawdown.min())


class ElectraTradingStrategy:
    """
    Trading strategy that uses ELECTRA sentiment for signal generation.

    The strategy:
    1. Analyzes news/text sentiment using ELECTRA
    2. Generates trading signals based on sentiment scores
    3. Sizes positions based on confidence levels
    4. Applies risk management rules

    Args:
        model: Trained FinancialElectra model
        vocab: Vocabulary dictionary for tokenization
        signal_threshold: Minimum absolute signal to trade
        confidence_threshold: Minimum confidence to trade
        max_position_size: Maximum position size as fraction of portfolio
        stop_loss_pct: Stop-loss as percentage of entry price
        holding_period: Maximum holding period in days
        device: Device for model inference
    """

    def __init__(
        self,
        model: FinancialElectra,
        vocab: Dict[str, int],
        signal_threshold: float = 0.3,
        confidence_threshold: float = 0.6,
        max_position_size: float = 1.0,
        stop_loss_pct: float = 0.05,
        holding_period: int = 5,
        device: str = 'cpu'
    ):
        self.signal_gen = ElectraTradingSignal(model, vocab, device)
        self.signal_threshold = signal_threshold
        self.confidence_threshold = confidence_threshold
        self.max_position_size = max_position_size
        self.stop_loss_pct = stop_loss_pct
        self.holding_period = holding_period

    def generate_signals(
        self,
        news_by_date: Dict[pd.Timestamp, List[str]]
    ) -> pd.DataFrame:
        """
        Generate trading signals from news data.

        Args:
            news_by_date: Dictionary mapping dates to lists of news texts

        Returns:
            DataFrame with columns: date, signal, confidence, direction, position_size
        """
        signals = []
        for date, news_items in sorted(news_by_date.items()):
            signal, confidence = self.signal_gen.generate_signal(
                news_items, threshold=self.confidence_threshold
            )

            if abs(signal) > self.signal_threshold:
                direction = 1 if signal > 0 else -1
                position_size = min(
                    abs(signal) * confidence,
                    self.max_position_size
                )

                signals.append({
                    'date': date,
                    'signal': signal,
                    'confidence': confidence,
                    'direction': direction,
                    'position_size': position_size
                })

        return pd.DataFrame(signals)


def run_backtest(
    strategy: ElectraTradingStrategy,
    price_data: pd.DataFrame,
    news_by_date: Dict[pd.Timestamp, List[str]],
    initial_capital: float = 100000.0
) -> BacktestResult:
    """
    Run a backtest of the ELECTRA trading strategy.

    Args:
        strategy: Configured ElectraTradingStrategy
        price_data: DataFrame with OHLCV price data
        news_by_date: Dictionary mapping dates to lists of news texts
        initial_capital: Starting portfolio value

    Returns:
        BacktestResult with performance metrics
    """
    signals = strategy.generate_signals(news_by_date)

    if signals.empty:
        equity = pd.Series(
            [initial_capital] * len(price_data),
            index=price_data.index
        )
        return BacktestResult(
            total_return=0.0,
            sharpe_ratio=0.0,
            sortino_ratio=0.0,
            max_drawdown=0.0,
            win_rate=0.0,
            profit_factor=0.0,
            num_trades=0,
            trades=[],
            equity_curve=equity,
            daily_returns=pd.Series(0.0, index=price_data.index)
        )

    trades = []
    capital = initial_capital
    equity_values = []
    daily_returns_list = []

    position = None
    position_days = 0

    for i, (date, row) in enumerate(price_data.iterrows()):
        # Check for exit conditions
        if position is not None:
            position_days += 1
            current_price = row['Close']

            # Check stop-loss
            if position['direction'] == 1:
                pnl_pct = (current_price - position['entry_price']) / position['entry_price']
            else:
                pnl_pct = (position['entry_price'] - current_price) / position['entry_price']

            should_exit = (
                pnl_pct <= -strategy.stop_loss_pct or
                position_days >= strategy.holding_period
            )

            if should_exit:
                pnl = pnl_pct * position['position_size'] * capital
                capital += pnl

                trade = Trade(
                    entry_date=position['entry_date'],
                    exit_date=date,
                    direction=position['direction'],
                    entry_price=position['entry_price'],
                    exit_price=current_price,
                    position_size=position['position_size'],
                    pnl=pnl,
                    pnl_pct=pnl_pct
                )
                trades.append(trade)
                position = None
                position_days = 0

        # Check for entry signals
        if position is None:
            matching_signals = signals[signals['date'] == date]
            if not matching_signals.empty:
                sig = matching_signals.iloc[0]
                position = {
                    'entry_date': date,
                    'entry_price': row['Close'],
                    'direction': int(sig['direction']),
                    'position_size': float(sig['position_size']),
                }
                position_days = 0

        equity_values.append(capital)
        if len(equity_values) > 1:
            daily_ret = (equity_values[-1] - equity_values[-2]) / equity_values[-2]
        else:
            daily_ret = 0.0
        daily_returns_list.append(daily_ret)

    equity_curve = pd.Series(equity_values, index=price_data.index)
    daily_returns = pd.Series(daily_returns_list, index=price_data.index)

    # Compute metrics
    total_return = (capital - initial_capital) / initial_capital
    sharpe = compute_sharpe_ratio(daily_returns)
    sortino = compute_sortino_ratio(daily_returns)
    max_dd = compute_max_drawdown(equity_curve)

    winning_trades = [t for t in trades if t.pnl > 0]
    losing_trades = [t for t in trades if t.pnl <= 0]
    win_rate = len(winning_trades) / len(trades) if trades else 0.0

    gross_profit = sum(t.pnl for t in winning_trades)
    gross_loss = abs(sum(t.pnl for t in losing_trades))
    profit_factor = gross_profit / gross_loss if gross_loss > 0 else float('inf')

    return BacktestResult(
        total_return=total_return,
        sharpe_ratio=sharpe,
        sortino_ratio=sortino,
        max_drawdown=max_dd,
        win_rate=win_rate,
        profit_factor=profit_factor,
        num_trades=len(trades),
        trades=trades,
        equity_curve=equity_curve,
        daily_returns=daily_returns
    )


def compare_strategies(
    results: Dict[str, BacktestResult]
) -> pd.DataFrame:
    """
    Compare performance of multiple strategy variants.

    Args:
        results: Dictionary mapping strategy names to BacktestResult objects

    Returns:
        DataFrame with comparison metrics
    """
    comparison = []
    for name, result in results.items():
        comparison.append({
            'Strategy': name,
            'Total Return': f"{result.total_return:.2%}",
            'Sharpe Ratio': f"{result.sharpe_ratio:.2f}",
            'Sortino Ratio': f"{result.sortino_ratio:.2f}",
            'Max Drawdown': f"{result.max_drawdown:.2%}",
            'Win Rate': f"{result.win_rate:.2%}",
            'Profit Factor': f"{result.profit_factor:.2f}",
            'Num Trades': result.num_trades,
        })

    return pd.DataFrame(comparison)

File: rust_examples/examples/backtest.rs
//! Example: Backtest ELECTRA sentiment trading strategy
//!
//! This example demonstrates backtesting a sentiment-based trading
//! strategy using ELECTRA models with Bybit cryptocurrency data.

use electra_finance::api::bybit::{BybitClient, Interval, Kline};
use electra_finance::data::processor::DataProcessor;
use electra_finance::nlp::tokenizer::SimpleTokenizer;
use electra_finance::models::network::SentimentNetwork;
use rand::Rng;

/// Generate synthetic news headlines based on price movements
fn generate_news_for_kline(kline: &Kline) -> Vec<String> {
    let ret = kline.return_pct();
    let pct = (ret.abs() * 100.0).max(0.1);

    if ret > 0.005 {
        vec![
            format!("Asset rallied {:.1}% on strong trading volume", pct),
            format!("Bullish momentum drove prices up {:.1}% today", pct),
        ]
    } else if ret < -0.005 {
        vec![
            format!("Asset declined {:.1}% amid heavy selling pressure", pct),
            format!("Bearish sentiment drove prices down {:.1}% today", pct),
        ]
    } else {
        vec![
            format!("Asset traded flat with minimal price change of {:.1}%", pct),
        ]
    }
}

/// Backtesting engine for sentiment-based strategies
struct Backtester {
    initial_capital: f64,
    signal_threshold: f64,
    stop_loss_pct: f64,
    holding_period: usize,
}

/// Single trade record
#[derive(Debug)]
struct TradeRecord {
    entry_idx: usize,
    exit_idx: usize,
    direction: i32,
    entry_price: f64,
    exit_price: f64,
    pnl_pct: f64,
}

impl Backtester {
    fn new(initial_capital: f64) -> Self {
        Self {
            initial_capital,
            signal_threshold: 0.1,
            stop_loss_pct: 0.05,
            holding_period: 5,
        }
    }

    fn run(
        &self,
        klines: &[Kline],
        signals: &[(f64, f64)], // (signal, confidence)
    ) -> BacktestResult {
        let mut capital = self.initial_capital;
        let mut trades: Vec<TradeRecord> = Vec::new();
        let mut equity_curve: Vec<f64> = Vec::with_capacity(klines.len());

        let mut position: Option<(usize, i32, f64)> = None; // (entry_idx, direction, entry_price)
        let mut days_held = 0usize;

        for i in 0..klines.len() {
            // Check exit conditions
            if let Some((entry_idx, direction, entry_price)) = position {
                days_held += 1;
                let current_price = klines[i].close;
                let pnl_pct = if direction == 1 {
                    (current_price - entry_price) / entry_price
                } else {
                    (entry_price - current_price) / entry_price
                };

                let should_exit = pnl_pct <= -self.stop_loss_pct
                    || days_held >= self.holding_period;

                if should_exit {
                    capital *= 1.0 + pnl_pct;
                    trades.push(TradeRecord {
                        entry_idx,
                        exit_idx: i,
                        direction,
                        entry_price,
                        exit_price: current_price,
                        pnl_pct,
                    });
                    position = None;
                    days_held = 0;
                }
            }

            // Check entry conditions
            if position.is_none() && i < signals.len() {
                let (signal, confidence) = signals[i];
                if signal.abs() > self.signal_threshold && confidence > 0.3 {
                    let direction = if signal > 0.0 { 1 } else { -1 };
                    position = Some((i, direction, klines[i].close));
                    days_held = 0;
                }
            }

            equity_curve.push(capital);
        }

        // Compute metrics
        let total_return = (capital - self.initial_capital) / self.initial_capital;
        let winning = trades.iter().filter(|t| t.pnl_pct > 0.0).count();
        let win_rate = if trades.is_empty() { 0.0 } else { winning as f64 / trades.len() as f64 };

        let gross_profit: f64 = trades.iter().filter(|t| t.pnl_pct > 0.0).map(|t| t.pnl_pct).sum();
        let gross_loss: f64 = trades.iter().filter(|t| t.pnl_pct <= 0.0).map(|t| t.pnl_pct.abs()).sum();
        let profit_factor = if gross_loss > 0.0 { gross_profit / gross_loss } else { f64::INFINITY };

        // Compute daily returns for Sharpe ratio
        let daily_returns: Vec<f64> = equity_curve
            .windows(2)
            .map(|w| (w[1] - w[0]) / w[0])
            .collect();

        let sharpe = compute_sharpe(&daily_returns);
        let sortino = compute_sortino(&daily_returns);
        let max_dd = compute_max_drawdown(&equity_curve);

        BacktestResult {
            total_return,
            sharpe_ratio: sharpe,
            sortino_ratio: sortino,
            max_drawdown: max_dd,
            win_rate,
            profit_factor,
            num_trades: trades.len(),
            final_capital: capital,
        }
    }
}

#[derive(Debug)]
struct BacktestResult {
    total_return: f64,
    sharpe_ratio: f64,
    sortino_ratio: f64,
    max_drawdown: f64,
    win_rate: f64,
    profit_factor: f64,
    num_trades: usize,
    final_capital: f64,
}

fn compute_sharpe(returns: &[f64]) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let n = returns.len() as f64;
    let mean = returns.iter().sum::<f64>() / n;
    let variance = returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / n;
    let std = variance.sqrt();
    if std == 0.0 { 0.0 } else { (252.0_f64).sqrt() * mean / std }
}

fn compute_sortino(returns: &[f64]) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let n = returns.len() as f64;
    let mean = returns.iter().sum::<f64>() / n;
    let downside: Vec<f64> = returns.iter().filter(|&&r| r < 0.0).cloned().collect();
    if downside.is_empty() {
        return 0.0;
    }
    let down_var = downside.iter().map(|r| r.powi(2)).sum::<f64>() / downside.len() as f64;
    let down_std = down_var.sqrt();
    if down_std == 0.0 { 0.0 } else { (252.0_f64).sqrt() * mean / down_std }
}

fn compute_max_drawdown(equity: &[f64]) -> f64 {
    let mut peak = f64::NEG_INFINITY;
    let mut max_dd = 0.0;

    for &val in equity {
        if val > peak {
            peak = val;
        }
        let dd = (val - peak) / peak;
        if dd < max_dd {
            max_dd = dd;
        }
    }

    max_dd
}

fn main() -> anyhow::Result<()> {
    env_logger::init();

    println!("=== ELECTRA Finance: Backtesting ===\n");

    // Create components
    let tokenizer = SimpleTokenizer::new(64);
    let model = SentimentNetwork::new(tokenizer.vocab_size(), 64, 128, 3);

    // Try fetching real data, fall back to synthetic
    let klines = match BybitClient::new().get_klines("BTCUSDT", Interval::Day1, 200, None, None) {
        Ok(k) if !k.is_empty() => {
            println!("Using real Bybit data: {} klines", k.len());
            k
        }
        _ => {
            println!("Generating synthetic data for demo...");
            generate_synthetic_klines(200)
        }
    };

    println!("Data points: {}\n", klines.len());

    // Generate sentiment signals for each kline
    let mut signals: Vec<(f64, f64)> = Vec::with_capacity(klines.len());

    for kline in &klines {
        let news = generate_news_for_kline(kline);
        let (batch_ids, batch_masks) = tokenizer.tokenize_batch(&news);
        let results = model.predict_batch(&batch_ids, &batch_masks);
        let (signal, confidence) = model.aggregate_signal(&results, 0.3);
        signals.push((signal, confidence));
    }

    // Run backtests with different configurations
    println!("--- Strategy 1: Default Parameters ---");
    let bt1 = Backtester::new(100_000.0);
    let result1 = bt1.run(&klines, &signals);
    print_result(&result1);

    println!("\n--- Strategy 2: Tight Stop-Loss ---");
    let mut bt2 = Backtester::new(100_000.0);
    bt2.stop_loss_pct = 0.02;
    bt2.holding_period = 3;
    let result2 = bt2.run(&klines, &signals);
    print_result(&result2);

    println!("\n--- Strategy 3: High Conviction Only ---");
    let mut bt3 = Backtester::new(100_000.0);
    bt3.signal_threshold = 0.3;
    bt3.holding_period = 7;
    let result3 = bt3.run(&klines, &signals);
    print_result(&result3);

    // Comparison
    println!("\n=== Strategy Comparison ===");
    println!("{:<25} {:>12} {:>10} {:>10} {:>10} {:>8}",
        "Strategy", "Return", "Sharpe", "Max DD", "Win Rate", "Trades");
    println!("{:-<85}", "");
    print_row("Default", &result1);
    print_row("Tight Stop-Loss", &result2);
    print_row("High Conviction", &result3);

    println!("\n=== Backtesting complete ===");

    Ok(())
}

fn print_result(result: &BacktestResult) {
    println!("  Total Return:   {:+.2}%", result.total_return * 100.0);
    println!("  Sharpe Ratio:   {:.2}", result.sharpe_ratio);
    println!("  Sortino Ratio:  {:.2}", result.sortino_ratio);
    println!("  Max Drawdown:   {:.2}%", result.max_drawdown * 100.0);
    println!("  Win Rate:       {:.1}%", result.win_rate * 100.0);
    println!("  Profit Factor:  {:.2}", result.profit_factor);
    println!("  Num Trades:     {}", result.num_trades);
    println!("  Final Capital:  ${:.2}", result.final_capital);
}

fn print_row(name: &str, result: &BacktestResult) {
    println!("{:<25} {:>+11.2}% {:>10.2} {:>+9.2}% {:>9.1}% {:>8}",
        name,
        result.total_return * 100.0,
        result.sharpe_ratio,
        result.max_drawdown * 100.0,
        result.win_rate * 100.0,
        result.num_trades
    );
}

/// Generate synthetic kline data for testing
fn generate_synthetic_klines(n: usize) -> Vec<Kline> {
    let mut rng = rand::thread_rng();
    let mut klines = Vec::with_capacity(n);
    let mut price = 40000.0; // Starting BTC price

    let start_ts = 1700000000000i64; // Nov 2023

    for i in 0..n {
        let change: f64 = rng.gen_range(-0.03..0.03);
        let open = price;
        let close = open * (1.0 + change);
        let high = open.max(close) * (1.0 + rng.gen_range(0.0..0.01));
        let low = open.min(close) * (1.0 - rng.gen_range(0.0..0.01));
        let volume = rng.gen_range(100.0..10000.0);

        klines.push(Kline {
            timestamp: start_ts + (i as i64 * 86400000),
            open,
            high,
            low,
            close,
            volume,
            turnover: volume * close,
        });

        price = close;
    }

    klines
}
